### DISCUSSION TITLE
Bug reports in RoboCup 2021

### DISCUSSION THREAD

**Discussion by TomohiroOno33** [hsr-project/general#20] [General]
RoboCup2021でのサーバ運営，ありがとうございました．
約1ヶ月の開発，テストを通して見つけたバグを報告します（システムの仕様もあるかもしれません，御了承ください）．
また，システムの改善案も考えましたので共有します．

# バグ
## 全体

1. シミュレータの起動について
#7 でも書きましたが，コントローラの起動に失敗することがあります．
ローカル・サーバどちらの場合においても失敗していることがありました．
シミュレータ立ち上げ後，起動（▷ボタンを押す）までに時間をおいても失敗していました．
（我々のローカル環境では非常に頻繁に起こっていました．3,4回に1回起動に成功するかどうかでした．）

2. Real Time Factorについて
別の場所で議論されたかと思いますが，RTFが高いと物体が滑りやすくなる，
また，離してから下に落ちるまでの時間が長くなる現象が確認されています．

3. 物体モデルについて
これはここで議論することではないかもしれませんが，いくつかの物体の衝突モデルが不完全なため，机や棚に埋まったり，手に刺さったりします．
特に，Tuna_fish_can，Plate，Cupsなどで顕著に見られました．

## Task 1

1. スコアが加算されないことがある．（特にDrawer left，Bin Aによく見られた）
2. スコアが二重で加算されることがある．（Drawer topに入れたときにしばしば起こる．おそらくtop, bottom両方の判定が入っている）
3. Collision判定が適用されないことがある．
4. Collision判定が間違って適用されることがある．
5. ルールにおいて，Collisionは−10点ではなく，*0.5である．
6. Drawerに多くの物体を入れると，衝突が起こってDrawerが移動したり，物体がDrawerから抜けたりする．
7. Binに多くの物体を入れると物体がBinから抜けることがある．（Drawerよりは低頻度）

## Task 2a

1. Task 1のあとにTask 2aを連続して行うと障害物を回避してもスコア（100点）が加算されないことがある．（おそらく，Task 1でのCollision判定を引きずっている）

# システム改善案
私一個人の意見です，参考までに．

## 全体

1. スコアUIのように，タイマーUIがあると良い．（特定のトピックを受け取ったら開始，などで管理する）
2. 衝突したときにUI上にメッセージが表示されるとわかりやすい． 
3. スコア増減のログが表示されていると何が起こっているか把握しやすい．
4. 結果発表のとき，スコアが見えている状態から始まるので少し面白味に欠ける．スコアのランクだけを表示し，スコア自体は隠しておくと見る楽しみが増えて良いと思う．

## Task 1

1. 引き出し開けをトピックでできるようにする．（例えば，/drawer_openのEmptyトピックが飛んできたら自動で開く，など．もちろん，ボーナスポイントは加算しない．）
2. （賛否あると思いますが）DrawerやBinに物体が入り，点数が加算されたあと，その物体を削除する．（こうすることで，衝突によるRTFの低下やバグでDrawerが移動したり，物体が抜けたりしなくなる）

**Comment by yosuke** [comment]
@TomohiroOno33 

> シミュレータの起動について
> SImulations fail on the scoring server #7 でも書きましたが，コントローラの起動に失敗することがあります．
> ローカル・サーバどちらの場合においても失敗していることがありました．
> シミュレータ立ち上げ後，起動（▷ボタンを押す）までに時間をおいても失敗していました．
>（我々のローカル環境では非常に頻繁に起こっていました．3,4回に1回起動に成功するかどうかでした．）

3,4回に1回起動だとかなり不便だったのではないかと思います。
OSS版HSRシミュレータは、全身逆運動学を使ったモーションの実行にfollow_trajectoryコントローラ、通常の移動時にcmd_velコントローラをros controllerの仕組みを使ってスイッチして動いています。
おそらくこのスイッチ部分が特定のタイミングで切り替わってくれていないのでは、、、と思うのですが、再現条件がわかったらより頑健に改善できればと思います。
rosnav移動とmoveitコマンドを同時利用するか短いタイミングで出すとロックする可能性もありそうですが、もし心当たりがありましたら教えてください。

> Real Time Factorについて
> 別の場所で議論されたかと思いますが，RTFが高いと物体が滑りやすくなる，
> また，離してから下に落ちるまでの時間が長くなる現象が確認されています．

これについても競技にかなり影響する部分だったと思うので、早めに改善したいです。
今回は、grasp-hackを使っていたのですが、物理エンジンレベルで滑りやすくなるというよりは、grasp-hackの物体のロック判定が遅くなる（同様にアンロックも遅くなる）現象だったのではないかと情報をいただいて考え始めました。
grasp-hackなしで把持シミュレーションを実現しようという話も出ており、その形での解決になるかも知れません。

> 物体モデルについて
> これはここで議論することではないかもしれませんが，いくつかの物体の衝突モデルが不完全なため，机や棚に埋まったり，手に刺さったりします．
> 特に，Tuna_fish_can，Plate，Cupsなどで顕著に見られました．

情報ありがとうございます。薄かったり、細かい構造があったりすると衝突判定して反力を加えるものの、強いトルクや自重が加わるとめり込んだり融合してしまったり、、、でおかしな挙動が出ます。
物理エンジンの根本的な性能によるところと、シミュレーション速度重視の設定によるところがあると思うのですが、なんとかしたいですね。

> Task 1
> スコアが加算されないことがある．（特にDrawer left，Bin Aによく見られた）
> スコアが二重で加算されることがある．（Drawer topに入れたときにしばしば起こる．おそらくtop, bottom両方の判定が入っている）

物体の中心位置が定義した領域に入っているかで判別しているのですが、Drawer topは引き出すと斜めに下がって下のDrawerに重なったりするので、ご推察の通り、それが原因になっていそうですね。

> Collision判定が適用されないことがある．
> Collision判定が間違って適用されることがある．
> ルールにおいて，Collisionは−10点ではなく，*0.5である．

ルールブック上では、Task1のCollisionのペナルティがあるのですが、判定が難しいことから今回については自動採点には組み込まれていませんでした。
モノをどこでも良いので片付けると10pt。カテゴリ毎に正しい場所に片付けるとプラスして10pt。ペンなどを正しい方向で片付けるとさらにプラスして10ptです。

> Drawerに多くの物体を入れると，衝突が起こってDrawerが移動したり，物体がDrawerから抜けたりする．
> Binに多くの物体を入れると物体がBinから抜けることがある．（Drawerよりは低頻度）

これも衝突判定の問題ですね。Drawerは結構薄いモデルになっていたので、もう少し厚く調整しても良かったかも知れません。

> Task 2a
> Task 1のあとにTask 2aを連続して行うと障害物を回避してもスコア（100点）が加算されないことがある．（おそらく，Task 1でのCollision判定を引きずっている）

Task1ではCollision判定していなかったので、別の原因かも知れません。
再現したら調査してみます。

> システム改善案
> 全体
> スコアUIのように，タイマーUIがあると良い．（特定のトピックを受け取ったら開始，などで管理する）
> 衝突したときにUI上にメッセージが表示されるとわかりやすい．
> スコア増減のログが表示されていると何が起こっているか把握しやすい．

コンソールにしかスコア関係のメッセージが表示されておらず、gazeboのwidgetとして色々表示できると良いですね。

> 結果発表のとき，スコアが見えている状態から始まるので少し面白味に欠ける．スコアのランクだけを表示し，スコア自体は隠しておくと見る楽しみが増えて良いと思う．

確かに。

> Task 1
> 引き出し開けをトピックでできるようにする．（例えば，/drawer_openのEmptyトピックが飛んできたら自動で開く，など．もちろん，ボーナスポイントは加算しない．）

リクエストはあったものの、実装が間に合わず、、、の機能でした。

> （賛否あると思いますが）DrawerやBinに物体が入り，点数が加算されたあと，その物体を削除する．（こうすることで，衝突によるRTFの低下やバグでDrawerが移動したり，物体が抜けたりしなくなる）

干渉形状の調整で解決できなかった場合は、これも現実的な解と思います。

**Reply by TomohiroOno33** [reply]
今後もこのシミュレータを使って開発をしていくつもりですので，少しでもお力になれたらと思います．

RTFに関しては，早いほうがもちろん嬉しいですが，安定していることが最も重要だと感じております．

> ルールブック上では、Task1のCollisionのペナルティがあるのですが、判定が難しいことから今回については自動採点には組み込まれていませんでした。
> モノをどこでも良いので片付けると10pt。カテゴリ毎に正しい場所に片付けるとプラスして10pt。ペンなどを正しい方向で片付けるとさらにプラスして10ptです。

たまに−10点の減点が入っている事がありましたが，これはCollisionによるペナルティだと思っていました．

> Task1ではCollision判定していなかったので、別の原因かも知れません。
> 再現したら調査してみます。

Task1実行時にもログ上ではCollision判定が流れていたので，そうかなと思っていました．

**Reply by yosuke** [reply]
こちらでも反省点を整理しているところだったので、バグや不具合のレポートはとてもありがたいです。

安定優先というご意見もごもっともと思います。

HSRと環境オブジェクトの衝突はここで検知していたのですが、設定上はtask1関連のオブジェクトは除外しており
https://github.com/hsr-project/tmc_gazebo_task_evaluators/blob/master/launch/robocup2021.launch#L209

自動採点のコードはこちらなのですが、task2a以外ではペナルティは設定していませんでした。
https://github.com/hsr-project/tmc_gazebo_task_evaluators/blob/master/src/wrs_score_counter.cpp

+10ptが記録された後でオブジェクトがボックスの範囲から外れたので0ptとなった（差分で-10ptになった）可能性もあるかも知れません。

**Reply by TomohiroOno33** [reply]
もしかしたら，机に当たっていたことが原因で減点されているかもしれないです．

**Reply by yosuke** [reply]
ただよくよく見るとbinが除外リストから抜けていますね、、、。

binに接触した場合は、task2aのペナルティが先に発生する形になるので、それで100ptが入らない場合があるということになっているかも知れません。

**Reply by yosuke** [reply]
@TomohiroOno33 確かに机も抜けていそう、、、。task1の減点にはアルゴリズム上はならないはずではあるのですが、task2aが影響を受ける可能性が大きいです。

**Reply by TomohiroOno33** [reply]
たまに机に当たっていたので（本当は良くないですが・・・），その可能性が高そうです．

> 物体の中心位置が定義した領域に入っているかで判別しているのですが、

物体の中心座標とは，モデルで定義されているものでしょうか？
（Gazebo上でモデルを動かすときに原点となっている点）
感覚ですが，下の方に設定されている物体が多かったので，物体の衝突判定の問題も相まって，点数が入らない事象が起きていたかもしれません．

**Reply by yosuke** [reply]
task2aについてはこのバグで間違いなさそうですね、、、。

競技に影響があったかも知れないことを心配しているのですが、大丈夫でしたでしょうか？
今回のRoboCupについては、task2aのオブジェクトを手で持って移動させていたチームもいたので、task2aについては目で見て確認もしていると思うのですが、、、。

**Reply by TomohiroOno33** [reply]
task2a（Go and Get it）の競技自体には問題なかったと思います．

私がこの問題に気がついたのは，Technical Challenge用のプログラムをサーバテストしたときで，時間もなかったのでその時はスルーしてました．
Technical Challengeはマニュアルで点数を動かしていないので，チームによっては影響が出ていたかもしれないです．
（少なくともうちのチームはtask 2aは0点でした）
ですが，そもそも家具にぶつかる事自体が問題なので，心配しなくても良いように思います・・・．

**Reply by yosuke** [reply]
> 物体の中心座標とは，モデルで定義されているものでしょうか？
> （Gazebo上でモデルを動かすときに原点となっている点）
> 感覚ですが，下の方に設定されている物体が多かったので，物体の衝突判定の問題も相まって，点数が入らない事象が起きていたかもしれません．

これもその感覚で正しい可能性が高いです。
判定に使っているのは、GazeboのAPIで取得できるオブジェクト位置、なのですが、XY軸方向では中心であるものの、Z軸では底に設定されているオブジェクトも多いようで、底が範囲からはみ出すと判定から外れてしまっていそうです。

**Reply by yosuke** [reply]
この状態だと、トレーにはみ出す姿勢で置くと、姿勢によっては範囲から出てしまうことが多くなりそうですね。
binも底の判定をギリギリにしていたかもしれないので、オブジェクトの底を下にして置くと判定がはみ出してしまうことがあったかも知れません。
boundingboxから中心座標を求め直した方が良いと思うので、うまくできるAPIがないか調べています。

**Reply by yosuke** [reply]
[メモ] GetLinkProperties APIを使うと重心（CoM）が取得できるので、これで補正するのが良いかも知れない。
http://docs.ros.org/en/jade/api/gazebo_msgs/html/srv/GetLinkProperties.html

**Reply by yosuke** [reply]
起動直後に動かない問題はこちらで解決中:
https://github.com/hsr-project/hsrb_robocup_dspl_docker/issues/2

**Comment by TomohiroOno33** [comment]
> task2aについてはこのバグで間違いなさそうですね、、、。
> 
> 競技に影響があったかも知れないことを心配しているのですが、大丈夫でしたでしょうか？ 今回のRoboCupについては、task2aのオブジェクトを手で持って移動させていたチームもいたので、task2aについては目で見て確認もしていると思うのですが、、、。

こちらのバグは，修正されていますでしょうか？
ローカル，サーバー共に，Task2aで点数が入らないバグが残っているように思います．
RCAPでは，Task1とTask2を連続して実行すると思いますので，影響が出る気がします．

**Reply by yosuke** [reply]
本件、対応が遅くなってしまいすいません。
直前になってしまいましたが、
- task1の机がコリジョン判定に入ってしまっており、task2aで点が入らない件
- task1のオブジェクトのいくつかの基準座標が底になっており、箱に入っても点が入らないことがある件
は修正予定です。
修正出来次第、連絡します。
